Auto-provision smart wallet

mhofman opened the pull request:
closes: #8558

## Description

This change implements smart wallet auto-provisioning by:
- Checking if a smart wallet is already provisioned in vstorage during admission checks
- Charging a new beans based fee if not
- Generating a provision action when processing the smart wallet message based on whether the smart wallet is provisioned already or not

It does not mark whether a smart wallet provision is pending, and as such may result in provision beans fees being charged multiple times if smart wallet messages are submitted and included in new blocks before the provision message is executed by swingset.

This PR includes some necessary changes to the upgrade handler:
- Generalizing the swingset parameters logic to always add new parameters using their default value in the upgrade handler

### Security Considerations

The provision step in swingset does not change and still provides a nominal amount of IST during provisioning, for each provision message handled. As such the beans amount charged for auto-provisioning a wallet should be higher than the nominal amount transferred from the provision pool.

### Scaling Considerations

None

### Documentation Considerations

TBD

### Testing Considerations

TBD

### Upgrade Considerations

Requires a new chain software consensus upgrade.
Added required logic to the upgrade handler for the new fee.

JimLarson commented:
High-level comments:

Impressive plumbing to get that all wired through!

I'm a little concerned about the additional complexity for trying to suppress duplicate auto-provisioning within the same transaction. We're already resigned to having duplicate auto-provision requests in different blocks while the first request is still queued, so maybe that's an acceptable way to handle the now-impossible case of multiple wallet messages in a single transaction? It would be great if we could drop the customization of `sdk.Context`.  Doing so would require both the ante handler and the msg server to check vstorage for the existence of the wallet, but I don't think that's prohibitive.

mhofman commented:
@JimLarson 

I removed the ctx logic. In hindsight, it's much better because it wouldn't be needed anyway if/when we implement marking of pending provisions. 

We could potentially re-use the egress marking of explicit provision to implement this marking now.

otoole-brendan commented:
https://github.com/Agoric/dapp-inter/issues/214 tracks the dapp-inter change to accommodate this

mhofman commented on golang/cosmos/x/swingset/types/msgs.go:
```diff
@@ -48,6 +51,52 @@ func chargeAdmission(ctx sdk.Context, keeper SwingSetKeeper, addr sdk.AccAddress
+	err = keeper.ChargeBeans(ctx, addr, beans)
+	if err != nil {
+		return ctx, err
+	}
```
> keeper.ChargeForSmartWallet

I was hoping to avoid yet another method on the `SwingSetKeeper` interface, but I suppose we'd need to have something to mark a pending smart wallet if/when we do that. It does feel a little weird to have some of the fee calculations in admission, some in the msg server, and now, some on the keeper as well.

michaelfig replied:
I don't really care where that method/function lives, just that it encapsulates the charging behaviour, so you could pass the keeper as an argument instead.  It can be a private helper function (starts with lowercase).

mhofman commented:
Addressed feedback, then rebased: [cf6f4d2...cdbaa77](https://github.com/Agoric/agoric-sdk/compare/cf6f4d250afb29f958256d4e18f692640d1bdf75..cdbaa774b64fdec04f7f3553d8a18157758625ae)

JimLarson commented on golang/cosmos/x/swingset/types/params.go:
```diff
@@ -163,3 +163,75 @@ func validateQueueMax(i interface{}) error {
 	return nil
 }
+
+// UpdateParams appends any missing params, configuring them to their defaults,
+// then returning the updated params or an error.
```
Add that existing params will not be modified, regardless of their value, and regardless of whether the parameter appears in the defaults.

JimLarson commented on golang/cosmos/vm/controller.go:
```diff
@@ -13,7 +13,11 @@ type ControllerContext struct {
+	// block or mempool. If no error is returned, it must return a context to use
+	// while processing the message. If no updated context is necessary it should
+	// return the input context.
+	CheckAdmissibility(sdk.Context, interface{}) (sdk.Context, error)
```
No implementation updates the context, so remove it from the return value.

JimLarson reviewed (changes requested):
Much simpler now! But a few remaining cleanups. Should be quick.

mhofman commented on golang/cosmos/x/swingset/keeper/keeper.go:
```diff
@@ -37,6 +37,12 @@ const (
 
+const (
+	// WalletStoragePathSegment matches the value of WALLET_STORAGE_PATH_SEGMENT
+	// packages/vats/src/core/startWalletFactory.js
```
I'll add a separate commit that I can skip when backporting to the release branch to avoid churn in bundles

mhofman commented on golang/cosmos/x/swingset/types/msgs.go:
```diff
@@ -48,6 +51,52 @@ func chargeAdmission(ctx sdk.Context, keeper SwingSetKeeper, addr sdk.AccAddress
+	// provision actions, and the provisioning code is not fully idempotent.
+	_, txShouldProvision := ctx.Context().Value(TxShouldProvision).(MsgProvision)
+	if txShouldProvision {
+		return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "cannot auto-provision for multiple smart wallet messages")
```
outdated

mhofman commented:
Rolled into #8623
